lib/modules/manager/composer/schema.ts

```diff
@@ -0,0 +1,313 @@
+import { z } from 'zod';
+import { logger } from '../../../logger';
+import { readLocalFile } from '../../../util/fs';
+import { regEx } from '../../../util/regex';
+import { Json, LooseArray, LooseRecord } from '../../../util/schema-utils';
+import { GitTagsDatasource } from '../../datasource/git-tags';
+import { GithubTagsDatasource } from '../../datasource/github-tags';
+import { PackagistDatasource } from '../../datasource/packagist';
+import { api as semverComposer } from '../../versioning/composer';
+import type { PackageDependency, PackageFileContent } from '../types';
+import type { ComposerManagerData } from './types';
+
+export const ComposerRepo = z.object({
+  type: z.literal('composer'),
+  /**
+   * The regUrl is expected to be a base URL. GitLab composer repository installation guide specifies
+   * to use a base URL containing packages.json. Composer still works in this scenario by determining
+   * whether to add / remove packages.json from the URL.
+   *
+   * See https://github.com/composer/composer/blob/750a92b4b7aecda0e5b2f9b963f1cb1421900675/src/Composer/Repository/ComposerRepository.php#L815
+   */
+  url: z.string().transform((url) => url.replace(/\/packages\.json$/, '')),
+});
+export type ComposerRepo = z.infer<typeof ComposerRepo>;
+
+export const GitRepo = z.object({
+  type: z.enum(['vcs', 'git']).transform(() => 'git' as const),
```

Comment:
wouldn't we always get git for VCS? isn't it over optimistic? 🤔

Reply:
It's makes current behavior more explicit. But if I implemented this manager from scratch, I would ignore the difference anyway. Seems like in practice they're same.